shell/platform/embedder/embedder.h

```diff
@@ -0,0 +1,102 @@
+// Copyright 2017 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#if defined(__cplusplus)
+extern "C" {
+#endif
+
+#ifndef FLUTTER_EXPORT
+#define FLUTTER_EXPORT
+#endif  // FLUTTER_EXPORT
+
+#define FLUTTER_ENGINE_VERSION 1
+
+typedef enum {
+  kSuccess = 0,
+  kInvalidLibraryVersion,
+  kInvalidArguments,
+} FlutterResult;
+
+typedef enum {
+  kOpenGL,
+} FlutterRendererType;
+
+typedef void* FlutterEngine;
+
+typedef bool (*BoolCallback)(void* /* user data */);
+typedef uint32_t (*UIntCallback)(void* /* user data */);
+
+typedef struct {
+  BoolCallback make_current;
+  BoolCallback clear_current;
+  BoolCallback present;
+  UIntCallback fbo_callback;
+} FlutterOpenGLRendererConfig;
+
+typedef struct {
+  FlutterRendererType type;
+  union {
+    FlutterOpenGLRendererConfig open_gl;
+  };
+} FlutterRendererConfig;
+
+typedef enum {
+  kWindowMetrics,
+  kPointer,
+} FlutterEventType;
+
+typedef struct {
+  size_t width;
+  size_t height;
+} FlutterWindowMetricsEvent;
+
+typedef enum {
+  kCancel,
+  kUp,
+  kDown,
+  kMove,
+} FlutterPointerPhase;
+
+typedef struct {
```

Comment:
> If we add fields to the struct, the implementation doesn't know whether the client was compiled before or after the fields were added.

I was going to use the `FLUTTER_ENGINE_VERSION` passed in during engine initialization to determine this. When we added to the struct, we would bump up the version. If the change included an ABI break, we don't initialize the engine at all. If there is a non-breaking update, we allow initialization but store the version to determine whether to read the added struct members. My hope was that this one mechanism will serve the dual purpose of addressing ABI breaking changes as well as non breaking updates.

But really, I don't have a particularly strong opinion about this. I will add the struct member. Do you want me to remove the `FLUTTER_ENGINE_VERSION` as well? This will box us into not allowing ABI breaking updates. But maybe that is a good thing?